From 6815d1d5dce25b2c571aa687254c2a8fdf31322a Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Mon, 04 Jan 2021 22:44:20 +0200
Subject: [PATCH] [SW-20719] pll: move code back to common

As a preparation for Gaudi requirement to scrub the HBM at preboot
stage, move the PLLs initialization code from goya2_arc_gnu back to
common place, where it can be used by any requiring user.

Change-Id: I6740275237c1466233356374ef1408b633fbac7b
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/pll.c b/common/pll.c
index 39e812f..d726254 100644
--- a/common/pll.c
+++ b/common/pll.c
@@ -82,8 +82,6 @@
 	}
 }
 
-__weak void hl_pll_init(struct hl_pll_cfg *hl_pll_cfg, int size) {}
-
 __weak u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
 {
 	u32 nr, nf, od, div_sel, factor, offset;
@@ -111,3 +109,117 @@
 
 	return out_freq;
 }
+
+static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
+{
+	regmap_writel(reset, cfg->base + HL_PLL_RST);
+}
+
+__weak int hl_pll_is_locked(struct hl_pll_cfg *cfg)
+{
+	return regmap_readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
+}
+
+__weak void hl_pll_cfg(struct hl_pll_cfg *cfg)
+{
+	hl_pll_reset(cfg, true);
+
+	regmap_writel(cfg->nr, cfg->base + HL_PLL_NR);
+	regmap_writel(cfg->nf, cfg->base + HL_PLL_NF);
+	regmap_writel(cfg->od, cfg->base + HL_PLL_OD);
+	regmap_writel(cfg->nb, cfg->base + HL_PLL_NB);
+	regmap_writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
+
+	/* after updating the coefficients we need to wait >=5us */
+	udelay(HL_PLL_MAX_WAIT_TIME);
+
+	hl_pll_reset(cfg, false);
+
+	/* Wait >500 (10us) reference cycles until PLL re-locks. */
+	udelay(HL_PLL_MAX_WAIT_TIME);
+}
+
+static void hl_pll_clk_switch_to_bypass(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u32 reg;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
+		regmap_writel(HL_PLL_REF_CLK, reg);
+	}
+}
+
+static void hl_pll_clk_switch_to_pll(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u32 reg;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
+		if (cfg->div_sel[sel] < 0 || cfg->div_sel[sel] > HL_PLL_DIV_PLL)
+			continue;
+		regmap_writel(cfg->div_sel[sel], reg);
+	}
+}
+
+static void hl_pll_div_init(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u32 div_base;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		if (cfg->div_sel[sel] == HL_PLL_REF_CLK ||
+		    cfg->div_factor[sel] < 0)
+			continue;
+
+		div_base = cfg->base + (sel * sizeof(u32));
+
+		regmap_writel(cfg->div_factor[sel],
+				div_base + HL_PLL_DIV_FACTOR);
+		regmap_writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
+		regmap_writel(1, div_base + HL_PLL_DIV_EN);
+	}
+
+	/* just to be on a safe side, wait for DIV init to propagate */
+	udelay(1);
+}
+
+static void hl_pll_wait_for_lock(struct hl_pll_cfg *cfg)
+{
+	int tout = HL_PLL_LOCK_TOUT;
+
+	while (!hl_pll_is_locked(cfg) && tout--)
+		udelay(1);
+
+	if (tout < 0) {
+		preboot_puts_raw("PLL lock timedout for");
+		preboot_puts(cfg->name);
+	}
+}
+
+static void hl_pll_set(struct hl_pll_cfg *cfg)
+{
+	hl_pll_clk_switch_to_bypass(cfg);
+
+	hl_pll_cfg(cfg);
+
+	/* Polling on PLL lock until lock is cleared or timed out */
+	hl_pll_wait_for_lock(cfg);
+
+	hl_pll_div_init(cfg);
+
+	hl_pll_clk_switch_to_pll(cfg);
+
+}
+
+/**
+ * hl_pll_init - Initialize pll registers
+ */
+void hl_pll_init(struct hl_pll_cfg *hl_pll_cfg, int size)
+{
+	int i;
+
+	for (i = 0; i < size; i++)
+		hl_pll_set(hl_pll_cfg + i);
+}
diff --git a/common/pll.h b/common/pll.h
index 6cca608..382c8f4 100644
--- a/common/pll.h
+++ b/common/pll.h
@@ -14,6 +14,13 @@
 #define SYS_REF_CLOCK		50000000ULL
 #define USEC_PER_SEC		1000000
 #define BYPASS_CYCLES_PER_USEC	(SYS_REF_CLOCK / USEC_PER_SEC)
+#define HL_PLL_MAX_WAIT_TIME	20	/* 20 us */
+
+#ifdef TARGET_IS_PLDM
+#define HL_PLL_LOCK_TOUT	10	/* 10 us */
+#else
+#define HL_PLL_LOCK_TOUT	5000	/* 5 ms */
+#endif
 
 enum hl_pll_bypass_val {
 	HL_PLL_REF_CLK,
diff --git a/goya2_arc_gnu/pll.c b/goya2_arc_gnu/pll.c
index 340c80c..5803eea 100644
--- a/goya2_arc_gnu/pll.c
+++ b/goya2_arc_gnu/pll.c
@@ -50,12 +50,10 @@
 }
 
 #ifdef TARGET_IS_PLDM
-/* PLL initializations for PLDM */
-#define HL_PLL_LOCK_TOUT	10	/* 10 us */
+
 #define GOYA2_PLL_CFG		0x100
 #define GOYA2_PLL_CFG_EN	BIT(0)
 #define GOYA2_PLL_CFG_LOCK	BIT(16)
-#define HL_PLL_DIV_FACTOR_CMD	0x220
 
 static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
 {
@@ -69,7 +67,7 @@
 	regmap_writel(reg, cfg->base + GOYA2_PLL_CFG);
 }
 
-static int hl_pll_is_locked(struct hl_pll_cfg *cfg)
+int hl_pll_is_locked(struct hl_pll_cfg *cfg)
 {
 	return regmap_readl(cfg->base + GOYA2_PLL_CFG) & GOYA2_PLL_CFG_LOCK;
 }
@@ -79,7 +77,7 @@
 #define GOYA2_PLL_REFDIV_SHIFT		24
 #define GOYA2_PLL_MAX_WAIT_TIME		1
 
-static void hl_pll_cfg(struct hl_pll_cfg *cfg)
+void hl_pll_cfg(struct hl_pll_cfg *cfg)
 {
 	u32 reg = 0;
 
@@ -102,91 +100,6 @@
 
 	/* enable the PLL by writing a value of 1 to CTRL_CFG.EN*/
 	hl_pll_reset(cfg, true);
-}
-
-static void hl_pll_clk_switch_to_bypass(struct hl_pll_cfg *cfg)
-{
-	int sel;
-	u32 reg;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
-		regmap_writel(HL_PLL_REF_CLK, reg);
-	}
-}
-
-static void hl_pll_clk_switch_to_pll(struct hl_pll_cfg *cfg)
-{
-	int sel;
-	u32 reg;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
-		if (cfg->div_sel[sel] < 0 || cfg->div_sel[sel] > HL_PLL_DIV_PLL)
-			continue;
-		regmap_writel(cfg->div_sel[sel], reg);
-	}
-}
-
-static void hl_pll_wait_for_lock(struct hl_pll_cfg *cfg)
-{
-	int tout = HL_PLL_LOCK_TOUT;
-
-	while (!hl_pll_is_locked(cfg) && tout--)
-		udelay(1);
-
-	if (tout < 0) {
-		preboot_puts_raw("PLL lock timedout for");
-		preboot_puts(cfg->name);
-	}
-}
-
-static void hl_pll_div_init(struct hl_pll_cfg *cfg)
-{
-	int sel;
-	u32 div_base;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		if (cfg->div_sel[sel] == HL_PLL_REF_CLK ||
-		    cfg->div_factor[sel] < 0)
-			continue;
-
-		div_base = cfg->base + (sel * sizeof(u32));
-
-		regmap_writel(cfg->div_factor[sel],
-				div_base + HL_PLL_DIV_FACTOR);
-		regmap_writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
-		regmap_writel(1, div_base + HL_PLL_DIV_EN);
-	}
-
-	/* just to be on a safe side, wait for DIV init to propagate */
-	udelay(1);
-}
-
-static void hl_pll_set(struct hl_pll_cfg *cfg)
-{
-	hl_pll_clk_switch_to_bypass(cfg);
-
-	hl_pll_cfg(cfg);
-
-	/* Polling on PLL lock until lock is cleared or timed out */
-	hl_pll_wait_for_lock(cfg);
-
-	hl_pll_div_init(cfg);
-
-	hl_pll_clk_switch_to_pll(cfg);
-
-}
-
-/**
- * hl_pll_init - Initialize pll registers
- */
-void hl_pll_init(struct hl_pll_cfg *hl_pll_cfg, int size)
-{
-	int i;
-
-	for (i = 0; i < size; i++)
-		hl_pll_set(hl_pll_cfg + i);
 }
 
 #endif /* TARGET_IS_PLDM */
